### DISCUSSION TITLE
Load configuration env file

### DISCUSSION THREAD

**Discussion by v41e** [aws/aws-pdk#666] [Q&A] [answered]
HI PDK community,

I'm trying to load env files to populate some CDK variables related to different environments (AWS account and AWS profile).

I looked at https://nx.dev/recipes/tips-n-tricks/define-environment-variables and it looks great but I don't understand why when i try to run a target with NX, it don't see the configuration part:

`pdk nx run infra:synth:dev` or `pdk nx run infra:synth --configuration dev`

In that scenario I want NX to load `.env.dev` (env specific vars), then `.env` (common vars) but i can't find the .env.dev
If I rename the `.env.dev` to `.env.synth`, this time it worked

if i add theses lines at the start of the `main.ts` in infra package

```
console.log(process.env.NX_TASK_TARGET_PROJECT)
console.log(process.env.NX_TASK_TARGET_TARGET)
console.log(process.env.NX_TASK_TARGET_CONFIGURATION)
```

I have output

```
infra
synth
undefined
```

Any idea with NX can't find the configuration ?

**Comment by mteichtahl** [comment]
Have you tried .env.synth.dev ?Please note: My working hours may not be your working hours. Please do not feel obligated to reply outside of your normal work schedule!Sent on the go - please excuse mistakes On 4 Dec 2023, at 8:18 pm, Vale ***@***.***> wrote:﻿
HI PDK community,
I'm trying to load env files to populate some CDK variables related to different environments (AWS account and AWS profile).
I looked at https://nx.dev/recipes/tips-n-tricks/define-environment-variables and it looks great but I don't understand why when i try to run a target with NX, it don't see the configuration part:
pdk nx run infra:synth:dev or pdk nx run infra:synth --configuration dev
In that scenario I want NX to load .env.dev (env specific vars), then .env (common vars) but i can't find the .env.dev
If I rename the .env.dev to .env.synth, this time it worked
if i add theses lines at the start of the main.ts in infra package
console.log(process.env.NX_TASK_TARGET_PROJECT)
console.log(process.env.NX_TASK_TARGET_TARGET)
console.log(process.env.NX_TASK_TARGET_CONFIGURATION)

I have output
infra
synth
undefined

Any idea with NX can't find the configuration ?

**Reply by v41e** [reply]
.env.synth.dev doesn't work too

**Comment by v41e** [comment]
I don't know if for the configuration to be recognize it hat to be in project.json, if yes how can i tell to nx monorepo that there are dev, test, stage and prod configurations in project.json, because there is no way to define configuration in the `MonorepoTsProject`

**Comment by agdimech** [comment]
Hi - before I dig deeper into this, have you seen this reference: https://nx.dev/recipes/tips-n-tricks/define-environment-variables ?

If you want configuration for a `dev` task, then you need a file called `.env.dev`. similarly for synth it would be `.env.synth`.

**Reply by v41e** [reply]
Hi @agdimech ,

Absolutely, this link is in the original question, that's why I think it's an out of the box feature.

The most simple way to reproduce is to try with a new monorepo:

1. Run `pdk new monorepo-ts`
2. Use a simple monorepo project with a simple TS app

```ts
import { MonorepoTsProject } from "@aws/pdk/monorepo";
import * as projen from "projen";

const monorepo = new MonorepoTsProject({
  devDeps: ["@aws/pdk"],
  name: "monorepo",
  projenrcTs: true,
});

new projen.typescript.TypeScriptAppProject({
  name: "app",
  outdir: "packages/app",
  parent: monorepo,
  defaultReleaseBranch: "main",
});

monorepo.synth();
```

3. Add in the hello.test.ts under app the following code

```ts
console.log(`NX_TASK_TARGET_PROJECT: ${process.env.NX_TASK_TARGET_PROJECT}`);
console.log(`NX_TASK_TARGET_TARGET: ${process.env.NX_TASK_TARGET_TARGET}`);
console.log(`NX_TASK_TARGET_CONFIGURATION: ${process.env.NX_TASK_TARGET_CONFIGURATION}`);
```

4. Run `pdk nx run app:build:dev`

According to official env vars https://nx.dev/reference/environment-variables the result should be:

```
NX_TASK_TARGET_PROJECT: app
NX_TASK_TARGET_TARGET: build
NX_TASK_TARGET_CONFIGURATION: dev
```

But i have:

```
NX_TASK_TARGET_PROJECT: app
NX_TASK_TARGET_TARGET: build
NX_TASK_TARGET_CONFIGURATION: undefined
```

The configuration part is not interpreted

**Reply by agdimech** [reply, marked as answer]
Gotcha,

The reason this is undefined is probably because you haven't specified a dev configuration for your build task. To do that you can do this:

```
import { MonorepoTsProject, NxProject } from "@aws/pdk/monorepo";
import * as projen from "projen";
import { NodePackageManager } from "projen/lib/javascript";

const monorepo = new MonorepoTsProject({
  devDeps: ["@aws/pdk"],
  name: "monorepo",
  packageManager: NodePackageManager.PNPM,
  projenrcTs: true,
});

const app = new projen.typescript.TypeScriptAppProject({
  name: "app",
  outdir: "packages/app",
  parent: monorepo,
  packageManager: NodePackageManager.PNPM,
  defaultReleaseBranch: "main",
});

NxProject.ensure(app).setTarget("build", {
  configurations: {
    dev: { },
  },
} as any); // have to cast to any as we don't currently expose the configurations prop

monorepo.synth();
```

**Ensure you run `pdk` after making the change to `.projenrc.ts`**

I then created a file in `packages/app/.env.build.dev` with `FOO=BAR` and modified `hello.test.ts` to `console.log(process.env.FOO);`.

Then when you run `pdk nx run app:build:dev` it will merge whatever values you have defined in the `dev` file (and object if you so choose) with a final output as follows:

```
  console.log
    NX_TASK_TARGET_PROJECT: app

      at Object.<anonymous> (test/hello.test.ts:5:11)
  console.log
    NX_TASK_TARGET_TARGET: build

      at Object.<anonymous> (test/hello.test.ts:6:11)
  console.log
    NX_TASK_TARGET_CONFIGURATION: dev

      at Object.<anonymous> (test/hello.test.ts:7:11)
  console.log
    FOO: BAR
```

**Reply by v41e** [reply]
Thank you so much! I was sure there was something missing in `nx.json` or `project.json` but I wasn't sure what! Thanks a lot, maybe a word on this in the documentation could help many other users, the dynamic env loading is really really useful

**Reply by agdimech** [reply]
You are welcome. Feel free to raise a request and we can add it to the backlog 🙂
